Add node marginalization support

hidmic opened the pull request:
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | Second half of #407 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | CPR's Jackal gazebo simulation [*] |

[*] See [gist](https://gist.github.com/hidmic/8bf35d9b4d4b8584ff26115f555eb787) for instructions on how to replicate it.

---

## Description of contribution in a few bullet points

* Exposed Ceres solver information matrix for covariance estimation of marginalized distributions. 
* Added node marginalization support, based on a Chow-Liu tree approximation to Karto's graph.
* Modified experimental lifelong mapping node to optionally use node marginalization (instead of plain removal).

## Description of documentation updates required from your changes

* Added new parameter, so I probably need to add that to the documentation page.

---

## Future work that may be required in bullet points

* This patch essentially reverts https://github.com/SteveMacenski/slam_toolbox/pull/445. I haven't run into issues myself, but I'm not entirely sure I follow what the failure mode was.
* Testing, lots of testing. This has been manually tested on a simulation only. It must be validated on a real mobile robot.
* Port to ROS 2. The only reason this patch targets ROS 1 Noetic is third-party simulation availability.

hidmic committed:
Add node marginalization support.

Based on a Chow-Liu tree approximation to Karto's graph
upon marginalization, using Karto's solver estimate of
the full joint information matrix to build new nonlinear
constraints.

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

SteveMacenski commented on slam_toolbox/lib/karto_sdk/include/karto_sdk/contrib/ChowLiuTreeApprox.h:
```diff
@@ -0,0 +1,54 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
```
This project is LGPL2.1, any contribution made should be made under these terms

hidmic replied:
Ah, I replicated SRI code license. We can certainly transition to LGPL v2.1. I'll review the terms first.

hidmic replied:
Changed in https://github.com/SteveMacenski/slam_toolbox/pull/468/commits/973f3b70dc16479107d51a5442aa08b58bd743eb.

SteveMacenski commented:
This is frankly so much code, I cannot reasonably review it. I'm not really sure what I can promise to be done here. There's enough going on and I don't have the time to do the background research to understand what you've done to critically review this. Is there someone that could take a look at this that would better understand and be able to provide technique and software reviews?

I've long since moved on from 2D SLAM into navigation world so this is also unfamiliar to me at this moment.

SteveMacenski commented:
This is frankly so much code, I cannot reasonably review it. I'm not really sure what I can promise to be done here. There's enough going on and I don't have the time to do the background research to understand what you've done to critically review this. Is there someone that could take a look at this that would better understand and be able to provide technique and software reviews?

hidmic commented:
> This is frankly so much code, I cannot reasonably review it. I'm not really sure what I can promise to be done here. 

Assuming that LOC count is the problem here, I think it would be possible to split this. At least to some extent.

> Is there someone that could take a look at this that would better understand and be able to provide technique and software reviews?

Hmm, good question. Perhaps I can ask folks over at Nav2's Slack? Or anybody at the Nav2 WG?

> I've long since moved on from 2D SLAM into navigation world so this is also unfamiliar to me at this moment.

That's fair. It is a bit intricate, that's true.

SteveMacenski commented:
It's not just LOC but also what's happening in those LOC. Though the length certainly doesn't help :laughing: 

No one that I can think of, there's not many people familiar with this codebase or the intricacies of SLAM, unfortunately.

hidmic committed:
Use LGPL 2.1 in lieu of LGPL 3

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

hidmic commented:
@SteveMacenski I've reached out to @Shokman to help review this contribution. He's a former co-worker and a friend with expertise in the area and no ties to Ekumen Inc. as of today. He has kindly accepted to do so. 

If you're onboard, I'll close this PR, split it in 2 or 3 smaller PRs, and kick start the review.

SteveMacenski commented on slam_toolbox/lib/karto_sdk/src/Mapper.cpp:
```diff
@@ -1659,6 +1664,40 @@ namespace karto
+    }
+
+    // see if edge already exists
+    for (Edge<LocalizedRangeScan> * pOtherEdge : GetEdges()) {
```
Iterating through the entire graph eek, is this necessary?

SteveMacenski replied:
Or can this be error checked before now not requiring the check here?

hidmic replied:
Hmm, yeah, this will degrade performance, and it's just a sanity check. Logic in `Mapper::MarginalizeNodeFromGraph()` below operates on the assumption that `edge already exists` will never be true when using this `MapperGraph::AddEdge` overload. 

Maybe we can put this loop within an `#ifdef _DEBUG` block. WDYT?

SteveMacenski replied:
If it should never be true, I would just remove it, or comment it out with a comment that it should never be true. Have you added a log and run this for a while and make sure you never see it (to also sanity check)

hidmic replied:
I have not, but I did compile for debug and there's an assertion below. It didn't ever abort (there would be a major logic error if it did).

SteveMacenski replied:
OK - please remove then. That would be nontrivial to deal with

hidmic replied:
Alright, dropped in https://github.com/SteveMacenski/slam_toolbox/pull/468/commits/c119840bef1eefd62e4735822723148913787f2e

SteveMacenski commented on slam_toolbox/solvers/ceres_solver.cpp:
```diff
@@ -37,7 +38,9 @@ CeresSolver::CeresSolver() :
 
   // formulate problem
-  angle_local_parameterization_ = AngleLocalParameterization::Create();
+  local_parameterization_ = new ceres::ProductParameterization(
```
Please explain this

hidmic replied:
I changed the formulation of the NLS problem to optimize SE3 variables instead of 3 scalar variables so as to maintain a 1-to-1 relationship between nodes and parameter blocks. Error terms are equivalent, but this way I can reconstruct the block ordering to make sense of the information matrix.

SteveMacenski replied:
https://github.com/ceres-solver/ceres-solver/blob/b0aef211db734379319c19c030e734d6e23436b0/docs/source/nnls_modeling.rst#classproductparameterization I noticed there is a depreciation warning, should we be using something else instead?

hidmic replied:
Hmm, I had not seen that. It looks like `Parameterization` abstractions have been deprecated in [Ceres 2.1.0rc1](https://github.com/ceres-solver/ceres-solver/blob/2.1.0rc1/docs/source/nnls_modeling.rst#classproductparameterization), but as far as I can see that version is not yet available in any Linux distro. Certainly not for Debian/Ubuntu, and [by the looks of it](https://tracker.debian.org/pkg/ceres-solver) I don't think it'll make it to Jammy.

We should probably keep an eye on it though, we may have to transition in the next couple years.

SteveMacenski commented on slam_toolbox/solvers/ceres_solver.cpp:
```diff
@@ -232,6 +233,31 @@ const karto::ScanSolver::IdPoseVector& CeresSolver::GetCorrections() const
+    std::vector<double*> parameter_blocks;
+    problem_->GetParameterBlocks(&parameter_blocks);
+    for (auto * block : parameter_blocks) {
+      (*ordering)[(*nodes_inverted_)[block]] = index;
```
Please explain this

hidmic replied:
I need a mechanism to recover the order of each node variable to make sense of the information matrix. When computing gradients and jacobians, Ceres matches the parameter block order (see [here](http://ceres-solver.org/nnls_modeling.html#_CPPv4N5ceres7Problem15EvaluateOptions16parameter_blocksE)). Parameter block order matches insertion order, until you [remove a block](https://github.com/ceres-solver/ceres-solver/blob/c8658c8992acd1a862c6a193b3f21e1ee0274344/internal/ceres/problem_impl.cc#L414-L439). Then that ordering is lost. Completely. This bit of code reconstructs that ordering, using a hashmap to recover the node ID for a given block.

SteveMacenski replied:
Is it not possible that we can combine the 2 maps with a struct so we have 1 data structure containing all of this information? I don't love the idea of having these 2 independent tables, it doesn't seem good

hidmic replied:
Hmm, what do you mean exactly? To build a bi-directional map of sorts backed by two hash maps? Or to have an actual bi-directional map?

hidmic replied:
Well, using a proper bidirectional map was a _tiny_ bit more troublesome than anticipated, as what I was doing here wasn't quite it (i.e. `Eigen::Vector3d` isn't `double *`). See https://github.com/SteveMacenski/slam_toolbox/pull/468/commits/1041a52eae3d3c75d1c48c87dedb3cae8bffb3d5. On the flip side, using a pool for parameter blocks may turn out to be beneficial to cache performance.

SteveMacenski replied:
Why not create a struct as the value in the map that contains more information than just the single entry?

hidmic replied:
> Why not create a struct as the value in the map that contains more information than just the single entry?

Hmm, I don't follow. The original hash map would map node ID to parameter block. I need to perform the reverse query. My previous approach used two hash maps. The current one uses a bidirectional map. Are you proposing to have a single hash map and perform a linear search every time we need to map a parameter block to a node ID? That's O(n (n + 1) / 2) in the worst case (i.e. next parameter block always found last).

SteveMacenski commented on slam_toolbox/solvers/ceres_solver.cpp:
```diff
@@ -37,7 +38,9 @@ CeresSolver::CeresSolver() :
 
   // formulate problem
-  angle_local_parameterization_ = AngleLocalParameterization::Create();
+  local_parameterization_ = new ceres::ProductParameterization(
```
This is `new`ed but never deleted in reset /  destructor

hidmic replied:
Ahh, good catch. Fixed in https://github.com/SteveMacenski/slam_toolbox/pull/468/commits/aa9ffa6f46a98b6972de9262a50cd45488973cfb.

hidmic replied:
BTW what's the point of dynamically allocating most `CeresSolver` member fields? Barring those that must be because Ceres' API requires it, it doesn't seem to add much value -- it only opens the door to leakage if constructor, destructor, and reset implementations ever end up out of sync.

SteveMacenski replied:
I think the things that are dynamically allocated are thoes that Ceres requires.

hidmic replied:
Hmm, those `std::unordered_map` members are not required by Ceres.

SteveMacenski replied:
I don't understand what you're referencing this comment is on `AngleLocalParameterization` which is a ceres-owned object . I personally don't like that I had it raw pointers and the expectation is that it deals with the lifecycle, but that's what Ceres docs said when I was looking at building this

hidmic replied:
> I don't understand what you're referencing this comment is on `AngleLocalParameterization`

Yes, you're right, my bad. I was referring to e.g. 
https://github.com/SteveMacenski/slam_toolbox/blob/646ef41b14e2b5f72e721963fafa490afa3d64fc/solvers/ceres_solver.hpp#L72-L73
Anyhow, it's orthogonal to this patch.

SteveMacenski commented on slam_toolbox/solvers/ceres_solver.cpp:
```diff
@@ -322,28 +357,18 @@ void CeresSolver::AddConstraint(karto::Edge<karto::LocalizedRangeScan>* pEdge)
-  sqrt_information(2,2) = precisionMatrix(2,2);
-
+  Eigen::Matrix3d sqrt_information =
+    pLinkInfo->GetCovariance().Inverse().ToEigen().llt().matrixL();
```
Did we want to use the new one from #473 ?

hidmic replied:
Yes! But I have to backport it to Noetic first.

SteveMacenski replied:
`Isbackported() {return true;};`

hidmic replied:
Indeed. I've rebased this PR since.

SteveMacenski commented on slam_toolbox/lib/karto_sdk/src/Mapper.cpp:
```diff
@@ -3017,38 +3056,112 @@ namespace karto
+              edge->GetSource()->GetObject()->GetUniqueId(),
+              edge->GetTarget()->GetObject()->GetUniqueId());
+          m_pGraph->RemoveEdge(edge);
+          delete edge;
```
Need to set to null so that any future check for `if(ptr)` will return `false`, that will not be done if not set to null ptr

hidmic replied:
Hmm, there are no future checks, and `edge` is scoped to the loop. Do you intend to set it to `nullptr` just to play it safe?

SteveMacenski replied:
Yeah, and there where places where this bit me in the ass badly when I was adding graph object removal and serialization. Anything deleted manually needs to have nullptr set in case something changes later. It needs to be possible to tell if a pointer is deleted or not. The only way is setting it to nullptr, just deleting it leaves it in limbo where `if(ptr)` checks don't return false. 

Take it from experience, forgetting this will cause nightmare level problems in the future :laughing:

hidmic replied:
Alright. Done in https://github.com/SteveMacenski/slam_toolbox/pull/468/commits/d65e03cb622e0e74c844be4adc55a0a5fc386a71.

hidmic committed:
Delete CeresSolver allocations upon destruction/reset

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

hidmic committed:
Drop duplicate edge checks from karto::Mapper

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

hidmic committed:
Rework parameter/residual block handling in Ceres solver

- Use a boost::bimap for node id <-> parameter block lookup
- Use a boost::pool for parameter block allocation
- Change Karto::Solver graph getter signature

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

hidmic committed:
Null edge pointer after delete

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>
